pkg/ottl/ottlfuncs/func_extract_pattern.go

```diff
@@ -0,0 +1,74 @@
+package ottlfuncs
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+type ExtractPatternArguments[K any] struct {
+	Target  ottl.StringGetter[K] `ottlarg:"0"`
+	Pattern string               `ottlarg:"1"`
+}
+
+func NewExtractPatternFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ExtractPattern", &ExtractPatternArguments[K]{}, createExtractPatternFunction[K])
```

Comment:
Since you can extract any number of patterns I think we should stick with a plural name: `ExtractPatterns`

Reply:
sounds good, updated.